server/src/main/java/org/opensearch/search/aggregations/bucket/filterrewrite/FilterRewriteOptimizationContext.java

```diff
@@ -123,15 +138,75 @@ public boolean tryOptimize(final LeafReaderContext leafCtx, final BiConsumer<Lon
         Ranges ranges = getRanges(leafCtx, segmentMatchAll);
         if (ranges == null) return false;
 
-        consumeDebugInfo(aggregatorBridge.tryOptimize(values, incrementDocCount, ranges));
+        if (hasSubAgg && this.segmentThreshold > leafCtx.reader().maxDoc() / ranges.getSize()) {
+            // comparing with a rough estimate of docs per range in this segment
+            return false;
+        }
+
+        Supplier<DocIdSetBuilder> disBuilderSupplier = getDocIdSetBuilderSupplier(leafCtx, values);
+        OptimizeResult optimizeResult;
+        try {
+            optimizeResult = aggregatorBridge.tryOptimize(values, incrementDocCount, ranges, disBuilderSupplier);
+            consumeDebugInfo(optimizeResult);
+        } catch (AbortFilterRewriteOptimizationException e) {
+            logger.error("Abort filter rewrite optimization, fall back to default path");
+            return false;
+        }
 
         optimizedSegments.incrementAndGet();
         logger.debug("Fast filter optimization applied to shard {} segment {}", shardId, leafCtx.ord);
         logger.debug("Crossed leaf nodes: {}, inner nodes: {}", leafNodeVisited, innerNodeVisited);
+        if (hasSubAgg) {
+            runSubAggCollections(leafCtx, collectableSubAggregators, optimizeResult.builders);
+        }
 
         return true;
     }
 
+    private static void runSubAggCollections(
+        LeafReaderContext leafCtx,
+        BucketCollector collectableSubAggregators,
+        DocIdSetBuilder[] builders
+    ) throws IOException {
+        LeafBucketCollector sub;
+        for (int bucketOrd = 0; bucketOrd < builders.length; bucketOrd++) {
+            logger.debug("Collecting bucket {} for sub aggregation", bucketOrd);
+            DocIdSetBuilder builder = builders[bucketOrd];
+            if (builder == null) {
+                continue;
+            }
+            DocIdSetIterator iterator = builders[bucketOrd].build().iterator();
+            sub = collectableSubAggregators.getLeafCollector(leafCtx); // resetting the sub collector after processing each bucket
+            while (iterator.nextDoc() != NO_MORE_DOCS) {
+                int currentDoc = iterator.docID();
+                sub.collect(currentDoc, bucketOrd);
+            }
```

Comment:
There is no need of doing this separately, since we are using the `MultiRangeTraversal` for collection only. The sub collectors should keep collecting the documents, as we keeping traversing through the tree.